### ISSUE TITLE
EMS calling point InsideHVACSystemIterationLoop is not effective for AirLoopHVAC

### ISSUE THREAD

**Issue by EnergyArchmage** [NREL/EnergyPlus#5521]
Originally From helpdesk ticket no. 11011. 

User noticed that EMS program was not behaving as expected.  Checking EDD output shows that sensor is being filled with a value that does not match the values of the output variable that the sensor is associated with.  

In EDD  files, this line shows the mistake. 
`RTU_4_BYPASS_FLOW_CONTROL,Line 10,SET RETURN_FLOW = RTU_4_RA_MASS_FLOW,16.614991, Occurrence info=RUN PERIOD 1, 07/02 12:42 - 12:43`

The value 16.614991 does not agree with the output variable.  It looks like the max mass flow rate, not the current flow rate.  RTU_4 doesn't control well because the sensor's data is not correct.  

need to check that GetInternalVariableValue is working correctly for EMS sensors. 

Originally submitted as version 8.3 file, but issue shows up in version 8.4 as well.  

Created a more reasonable file derived from reference building large office file.  
### Checklist
- [x] Defect file added (https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/5000s/5521)
- [ ] Ticket added to Pivotal for defect (development team task)
- [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by rraustad** [comment]
@DeadParrot what would a user do when an input file does not conform to this assert? Should the code have allowed this input file to ever get here?

line 595 DataSurfaces:
assert( ( shapeCat == ShapeCat::Nonconvex ) || ( crossEdges.size() == 2 ) );

here crossEdges.size = 4.

**Comment by DeadParrot** [comment]
@rraustad the assert is saying that either a surface should be flagged as nonconvex or else it should have exactly 2 edges that cross a y=constant line between the bottom and top vertex y positions. If you show me the surface in question or the idf file in question and the revision(s) where the problem appears I can take a look and see what is wrong.

**Comment by rraustad** [comment]
I'm running the defect file for this issue.

**Comment by rraustad** [comment]
I think the first surface that failed is Surface 784.

**Comment by DeadParrot** [comment]
@rraustad how do I get the defect file? I'm not registered on the HelpDesk at this point.

**Comment by EnergyArchmage** [comment]
File is here

https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/5000s/5521

**Comment by rraustad** [comment]
This file has max iteration warnings and I wonder if these don't line up when max iter occurs? I saw a lot of times when max iter was occurring and only see a few times when data doesn't line up so not sure there is a correlation there or not.

EMS data is from NotePad++ where I searched for:
RTU_4_BYPASS_FLOW_CONTROL,Line 10,SET RETURN_FLOW = RTU_4_RA_MASS_FLOW
and then copied the data out of the lower window to excel. Then deleted repetitive (max iter) data.

![emsdata_5521](https://cloud.githubusercontent.com/assets/5778156/13678210/8530436e-e6bb-11e5-829e-9b13566f1670.png)

**Comment by DeadParrot** [comment]
@rraustad The idf has at least 2 nonconvex surfaces: 355 and 784. They are not being marked as nonconvex due to the bug in #5507 so the assert triggers.

To fix #5507 it should suffice to move Polygonal to the end of the list and give it a value of 10 (although making it an enum would be nicer). It may be that @mjwitte is going to do this fix soon.

**Comment by mjwitte** [comment]
@rraustad I was hoping to work on this fix soon, but feel free to do that fix - should probably be a separate PR but we can expedite getting it merged.

**Comment by rraustad** [comment]
Polygonal(10) did the trick, this file now runs in the debugger.

**Comment by EnergyArchmage** [comment]
Back to the original issue, I just want to note that there are no instance in the testfiles suite where file has both InsideHVACSystemIterationLoop and a sensor that is placed on a variable that is system HVAC timestep.  Smells like a clue to me.

**Comment by EnergyArchmage** [comment]
Developed a more reasonable defect file based on reference building large office and uploaded to defect files area.   Testing and running in debugger shows no real problem with EMS or sensor filling, the Node.MassFlowRate values really do have the erroneous values inside the system iterations.  So it looks like an air system mass flow rate modeling problem.  For calling point InsideHVACSystemIterationLoop, the errors show in EMS sensor values, but the correct mass flow rates are present for all other calling points.  Here is a snippet of EDD that shows the problem in the new defect file. For some reason the branch's inlet node is okay(5.46 kg/s), but not the downstream nodes (18.2 kg/s). 

```
TEST_INSIDEHVACSYSTEMITERATIONLOOP,Line 1,SET DUMMY = NODE_MDOT_EQUIPINLET,5.460051, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
TEST_INSIDEHVACSYSTEMITERATIONLOOP,Line 2,SET DUMMY = NODE_MDOT_MA,18.200171, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
TEST_INSIDEHVACSYSTEMITERATIONLOOP,Line 3,SET DUMMY = NODE_MDOT_CCOUT,18.200171, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
TEST_INSIDEHVACSYSTEMITERATIONLOOP,Line 4,SET DUMMY = NODE_MDOT_HCOUT,18.200171, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
TEST_INSIDEHVACSYSTEMITERATIONLOOP,Line 5,SET DUMMY = NODE_MDOT_FANOUT,18.200171, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
TEST_ENDOFSYSTEMTIMESTEPBEFOREHVACREPORTING,Line 1,SET DUMMY = NODE_MDOT_EQUIPINLET,5.460051, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
TEST_ENDOFSYSTEMTIMESTEPBEFOREHVACREPORTING,Line 2,SET DUMMY = NODE_MDOT_MA,5.460051, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
TEST_ENDOFSYSTEMTIMESTEPBEFOREHVACREPORTING,Line 3,SET DUMMY = NODE_MDOT_CCOUT,5.460051, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
TEST_ENDOFSYSTEMTIMESTEPBEFOREHVACREPORTING,Line 4,SET DUMMY = NODE_MDOT_HCOUT,5.460051, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
TEST_ENDOFSYSTEMTIMESTEPBEFOREHVACREPORTING,Line 5,SET DUMMY = NODE_MDOT_FANOUT,5.460051, During Warmup, Occurrence info=MAY2, 05/02 05:50 - 06:00
```

**Comment by rraustad** [comment]
What was the flow rate on the previous iteration?

**Comment by EnergyArchmage** [comment]
It ended with the 5.460051 kg/s value.  The 18.2 looks like the max.

**Comment by mjwitte** [comment]
A couple of thoughts.  On FirstHVACIteration, the maxavails are set to full flow and it takes a couple of iterations up and back down the zonesplitter (or plenum) to communicate that back to the branch with the fan.  Also, the OA mixer can alter the mixed flow rate.

**Comment by EnergyArchmage** [comment]
A few other findings:  
- same behavior in V8.1
- did not find issue occuring for plant nodes.

**Comment by mjwitte** [comment]
And if you're using current develop, it's got the extra reset on airflows back to max, every five iterations.  #5159 and #5532 both remove [this line](https://github.com/NREL/EnergyPlus/pull/5159/files#diff-cd290070b0d136de4179479f221302dbR1622) and should be merged in soon.

**Comment by EnergyArchmage** [comment]
Okay, changing how the air loop solver works is not in scope for this issue.  It seems that all that is needed to get EMS to work for this calling point is to make sure there is a second iteration with FirstHVACIteration set false.  This way the EMS controls can have accurate sensors and be effective for the second iteration.  Testing a fix on this branch.

https://github.com/NREL/EnergyPlus/tree/5521-EMS-InsideHVACSystemIterationLoop-AirLoop

**Comment by Myoldmopar** [comment]
Closed via #5564
